## <b> problem statement </b>
- <b> <i> FoodHunter </i> </b> , a leading food delivery app, has experienced a drop in revenue over the four months from June 2022 to September 2022. The goal is to identify the causes and contributing factors behind this decline and develop strategies to overcome these challenges.

## <b> loading the SQL extension and connecting to the <i> foodhunter database </i> </b>

In [2]:
%load_ext sql
%sql mysql+mysqlconnector://root:root@localhost/foodhunter

In [3]:
%%sql
show tables;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
10 rows affected.


[('avg_deliverytime_and_revenue',),
 ('customers',),
 ('dining_segment_summary',),
 ('drivers',),
 ('food_items',),
 ('fooditems_preferences',),
 ('orders',),
 ('orders_items',),
 ('orders_ordersitems_fooditems',),
 ('restaurants',)]

## <b> Investigating tables</b>

### <b> orders</b>

In [3]:
%%sql
select
column_name from information_schema.columns
where
table_schema = "foodhunter" and table_name = "orders";

 * mysql+mysqlconnector://root:***@localhost/foodhunter
13 rows affected.


[('customer_id',),
 ('delivered_date',),
 ('delivered_time',),
 ('delivery_address',),
 ('delivery_fee',),
 ('discount',),
 ('driver_id',),
 ('final_price',),
 ('order_date',),
 ('order_id',),
 ('order_rating',),
 ('order_time',),
 ('total_price',)]

In [4]:
%%sql
desc foodhunter.orders;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
13 rows affected.


[('order_id', 'int', 'YES', '', None, ''),
 ('customer_id', 'bigint', 'YES', '', None, ''),
 ('order_date', 'datetime', 'YES', '', None, ''),
 ('order_time', 'time', 'YES', '', None, ''),
 ('delivered_date', 'text', 'YES', '', None, ''),
 ('delivered_time', 'time', 'YES', '', None, ''),
 ('delivery_address', 'text', 'YES', '', None, ''),
 ('driver_id', 'bigint', 'YES', '', None, ''),
 ('total_price', 'bigint', 'YES', '', None, ''),
 ('delivery_fee', 'double', 'YES', '', None, ''),
 ('discount', 'double', 'YES', '', None, ''),
 ('final_price', 'double', 'YES', '', None, ''),
 ('order_rating', 'double', 'YES', '', None, '')]

In [5]:
%%sql
select min(order_date) as first_date_orders, 
max(order_date) as last_date_orders 
from foodhunter.orders;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
1 rows affected.


[(datetime.datetime(2022, 6, 1, 0, 0), datetime.datetime(2022, 9, 29, 0, 0))]

### observations
* We are analyzing data from 1st June 2022 to 29th September 2022 to understand why there was a drop in revenue for foodhunter during this period and to identify the underlying causes.

### <b> orders_items</b>

In [6]:
%%sql
select
column_name from information_schema.columns
where
table_schema = "foodhunter" and table_name = "orders_items";

 * mysql+mysqlconnector://root:***@localhost/foodhunter
3 rows affected.


[('item_id',), ('order_id',), ('quantity',)]

In [7]:
%%sql
desc foodhunter.orders_items;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
3 rows affected.


[('order_id', 'int', 'YES', '', None, ''),
 ('item_id', 'int', 'YES', '', None, ''),
 ('quantity', 'int', 'YES', '', None, '')]

### <b> restaurants</b>

In [8]:
%%sql
select
column_name from information_schema.columns
where
table_schema = "foodhunter" and table_name = "restaurants";

 * mysql+mysqlconnector://root:***@localhost/foodhunter
4 rows affected.


[('address',), ('cuisine',), ('restaurant_id',), ('restaurant_name',)]

In [9]:
%%sql
desc foodhunter.restaurants;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
4 rows affected.


[('restaurant_id', 'int', 'NO', 'PRI', None, ''),
 ('restaurant_name', 'text', 'YES', '', None, ''),
 ('cuisine', 'text', 'YES', '', None, ''),
 ('address', 'text', 'YES', '', None, '')]

### <b> customers</b>

In [10]:
%%sql
select
column_name from information_schema.columns
where
table_schema = "foodhunter" and table_name = "customers";

 * mysql+mysqlconnector://root:***@localhost/foodhunter
5 rows affected.


[('address',),
 ('customer_id',),
 ('first_name',),
 ('last_name',),
 ('phone_number',)]

In [11]:
%%sql
desc foodhunter.customers;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
5 rows affected.


[('customer_id', 'int', 'NO', 'PRI', None, ''),
 ('first_name', 'text', 'YES', '', None, ''),
 ('last_name', 'text', 'YES', '', None, ''),
 ('phone_number', 'text', 'YES', '', None, ''),
 ('address', 'text', 'YES', '', None, '')]

### <b> drivers</b>

In [12]:
%%sql
select
column_name from information_schema.columns
where
table_schema = "foodhunter" and table_name = "drivers";

 * mysql+mysqlconnector://root:***@localhost/foodhunter
3 rows affected.


[('driver_id',), ('name',), ('rating',)]

In [13]:
%%sql
desc foodhunter.drivers;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
3 rows affected.


[('driver_id', 'int', 'NO', 'PRI', None, ''),
 ('name', 'text', 'YES', '', None, ''),
 ('rating', 'double', 'YES', '', None, '')]

### <b> food_items</b>

In [14]:
%%sql
select
column_name from information_schema.columns
where
table_schema = "foodhunter" and table_name = "food_items";

 * mysql+mysqlconnector://root:***@localhost/foodhunter
6 rows affected.


[('calories',),
 ('food_type',),
 ('item_id',),
 ('item_name',),
 ('price',),
 ('restaurant_id',)]

In [15]:
%%sql
desc foodhunter.food_items;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
6 rows affected.


[('item_id', 'int', 'NO', 'PRI', None, ''),
 ('restaurant_id', 'int', 'YES', 'MUL', None, ''),
 ('item_name', 'text', 'YES', '', None, ''),
 ('food_type', 'text', 'YES', '', None, ''),
 ('price', 'text', 'YES', '', None, ''),
 ('calories', 'double', 'YES', '', None, '')]

### <b> The rate of change in order count and revenue over the months </b>

In [16]:
%%sql
with order_revenue_summary as (
  select 
  concat(year(order_date), '-', month(order_date)) as year__month, 
  count(distinct order_id) as ordercount, 
  round(sum(final_price), 2) as tot_revenue 
  from orders 
  group by concat(year(order_date), '-', month(order_date))
),
lag_order_revenue_summary as(
  select
  year__month,
  ordercount,
  lag(ordercount) over(order by year__month asc) as lag_order_count,
  tot_revenue,
  lag(tot_revenue) over(order by year__month asc) as lag_tot_revenue
  from order_revenue_summary
)

select 
  year__month,
  ordercount,
  lag_order_count,
  round(((ordercount - lag_order_count) / lag_order_count) * 100, 2) as ordercount_change_in_percent,
  tot_revenue,
  lag_tot_revenue,
  round(((tot_revenue - lag_tot_revenue) / lag_tot_revenue) * 100, 2) as tot_revenue_change_in_percent
from lag_order_revenue_summary;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
4 rows affected.


[('2022-6', 12502, None, None, 347577.5, None, None),
 ('2022-7', 11144, 12502, Decimal('-10.86'), 308601.5, 347577.5, -11.21),
 ('2022-8', 10107, 11144, Decimal('-9.31'), 283365.9, 308601.5, -8.18),
 ('2022-9', 9365, 10107, Decimal('-7.34'), 258161.1, 283365.9, -8.89)]

#### <b> observations </b>
* From the data, we observe a downward trend in order count and total revenue from June 2022 to September 2022.
* <b> The order count </b> shows a 10.86% drop from June to July, 9.31% drop from July to August and 7.34% drop from August to September. While the rate of decline is slowing, the overall drop in order count is still significant and requires attention.
* On the Otherhand, <b> The total revenue </b> shows an 11.21% drop from June to July, 8.18% drop from July to August and 8.89% drop from August to September. Although the rate of decline fluctuates, the overall decrease in total revenue is significant and requires immediate attention.
* Based on the rate of change, we can infer that the factors causing the decline are fading. It seems that recovering phase gets started or marketing strategies have started to adjust and adapt, leading to a reduction in the impact of losses.
* It is important to analyze the data at various granularities, as revenue shows signs of volatility in its rate of change.

### <b> Weekday vs Weekend: Monthly Variations in Orders and Revenue </b>

In [17]:
%%sql
select
  weekpart,
  count(distinct order_id) as ordercount,
  round(sum(final_price), 2) as tot_revenue,
  case when weekpart = 'weekday' 
    then count(distinct order_id) / 5
    else count(distinct order_id) / 2
  end as ordercount_perday_approx,
  case when weekpart = 'weekday' 
    then round(sum(final_price), 2) / 5
    else round(sum(final_price), 2) / 2
  end as tot_revenue_perday_approx
from
  (select
    *,
    case 
     when dayofweek(order_date) in (1,7) then 'weekend' 
     else 'weekday' 
    end as weekpart
  from orders) q1
group by weekpart;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
2 rows affected.


[('weekday', 31412, 870423.1, Decimal('6282.4000'), 174084.62),
 ('weekend', 11706, 327282.9, Decimal('5853.0000'), 163641.45)]

#### <b> observations </b>
* From this we can conclude, Weekdays have higher revenue and order count compared to weekends.

In [18]:
%%sql

with weekpart_order_revenue_summary as(
  select
  concat(year(order_date), '-', month(order_date)) as year__month,
  weekpart,
  sum(ordercount) as ordercount,
  round(sum(tot_revenue), 2) as tot_revenue
  from
   (select
     order_date,
     case
       when dayofweek(order_date) in (1,7) then "weekend"
       else "weekday"
     end as weekpart,
     count(distinct order_id) as ordercount, 
     sum(final_price) as tot_revenue 
    from orders
     group by order_date) q1
  group by concat(year(order_date), '-', month(order_date)), weekpart
),
lag_weekpart_summary as(
  select
    year__month,
    weekpart,
    ordercount,
    lag(ordercount) over(partition by weekpart order by year__month) as lag_ordercount,
    tot_revenue,
    lag(tot_revenue) over(partition by weekpart order by year__month) as lag_tot_revenue
  from weekpart_order_revenue_summary
)

select 
  year__month,
  weekpart,
  ordercount,
  round(((ordercount - lag_ordercount) / lag_ordercount)*100, 2) as ordercount_change_in_percent,
  tot_revenue,
  round(((tot_revenue - lag_tot_revenue) / lag_tot_revenue)*100, 2) as tot_revenue_change_in_percent
from lag_weekpart_summary;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
8 rows affected.


[('2022-6', 'weekday', Decimal('9043'), None, 250657.7, None),
 ('2022-7', 'weekday', Decimal('8096'), Decimal('-10.47'), 224993.4, -10.24),
 ('2022-8', 'weekday', Decimal('7384'), Decimal('-8.79'), 205586.2, -8.63),
 ('2022-9', 'weekday', Decimal('6889'), Decimal('-6.70'), 189185.8, -7.98),
 ('2022-6', 'weekend', Decimal('3459'), None, 96919.8, None),
 ('2022-7', 'weekend', Decimal('3048'), Decimal('-11.88'), 83608.1, -13.73),
 ('2022-8', 'weekend', Decimal('2723'), Decimal('-10.66'), 77779.7, -6.97),
 ('2022-9', 'weekend', Decimal('2476'), Decimal('-9.07'), 68975.3, -11.32)]

#### <b> observations </b>
* On both weekdays and weekends, order count and total revenue show a downward trend from June to September, with the rate of decrease gradually slowing over the months.
* For weekend revenues, the rate of decline fluctuates dropping from 13% (from june to july) to 6% (from july to august) indicating a slowdown followed by a huge drop of 11.32% from August to September.
* From this observation at this level of granularity, we can conclude that revenue changes across the four months for weekends are highly volatile (with a fluctuations in the rate of change), while the rate of change in order count remains relatively declines.
* For Weekdays, orders and total revenue show less volatility in their rate of change, the drops are becoming less steep over time.
* FoodHunter should focus more on weekend operations than weekday operations due to revenue fluctuations, as the current strategy does not seem effective for weekends in terms of revenue.
* Weekend operations should be prioritized as the primary focus due to significant revenue fluctuations with weekday operations addressed as the next level of importance.

### <b> OrderCount & Total Revenue Variation Across Time Segments </b>

#### <b> <i> Segmenting time periods according to customer food habits </i></b>
- #### <b> <i> order_timings => dining_segment </i></b>
- 07:00 – 10:00	=> Breakfast
- 10:01 – 11:59	=> Brunch
- 12:00 – 14:59	=> Lunch
- 15:00 – 18:59	=> Snacks
- 19:00 – 22:00	=> Dinner
- 22:01 – 06:59	=> After hours

In [19]:
%%sql

drop view if exists dining_segment_summary;


create view dining_segment_summary as
select
  year__month,
  weekpart,
  dining_segment,
  count(distinct order_id) as ordercount,
  sum(final_price) as tot_revenue
from
  (select
     concat(year(order_date), "-", month(order_date)) as year__month,
     case
        when dayofweek(order_date) in (1,7) then "weekend"
        else "weekday"
     end as weekpart,
     case
        when order_time between '07:00:00' and '10:00:00' 
        then '1-Breakfast'
        when order_time between '10:01:00' and '11:59:59' 
        then '2-Brunch'
        when order_time between '12:00:00' and '14:59:59' 
        then '3-Lunch'
        when order_time between '15:00:00' and '18:59:59' 
        then '4-Snacks'
        when order_time between '19:00:00' and '22:00:00' 
        then '5-Dinner'
        when order_time between '22:01:00' and '23:59:59' or order_time between '00:00:00' and '06:59:59'  
        then '6-After hours'
     end as dining_segment,
     order_id,
     final_price
  from orders) q1
group by year__month, weekpart, dining_segment
order by year__month;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
0 rows affected.
0 rows affected.


[]

In [20]:
%%sql

select
  dining_segment,
  sum(ordercount) as ordercount,
  round(sum(tot_revenue), 2) as tot_revenue
from dining_segment_summary
group by dining_segment
order by tot_revenue desc, ordercount desc;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
6 rows affected.


[('6-After hours', Decimal('13726'), 380464.4),
 ('4-Snacks', Decimal('7797'), 215378.5),
 ('3-Lunch', Decimal('5950'), 167552.6),
 ('1-Breakfast', Decimal('5990'), 166993.6),
 ('5-Dinner', Decimal('5890'), 162552.1),
 ('2-Brunch', Decimal('3765'), 104764.8)]

#### <b> observations </b>
* The After Hours segment (10 PM to 7 AM) records the highest number of orders and total revenue, which could be attributed to several factors Limited restaurant availability, restricted mobility, Late-night cravings, social events and irregular sleep schedules also contribute to higher demand.
* Brunch ranks the lowest in both order count and total revenue possibly due to it not aligning with the typical eating habits of most people.
* To retain customers across different time segments, continuous offers and engagement is needed.
* Delays can significantly impact customer satisfaction and lead to a decrease in revenue.
* To prevent delays and resource shortages, delivery partners should be strategically allocated based on time periods, especially during After Hours and evening times. They should be on standby during peak hours and can take rest or go offline during periods of low demand or normal demand
* To increase the customer base during morning and lunch hours, we can introduce a variety of cuisines especially, those that align with local preferences. Offering regionally popular options could boost both orders and revenue.

In [21]:
%%sql

with weekdaylag_dining_segment_summary as (
  select 
    year__month,
    dining_segment,
    ordercount,
    lag(ordercount) over(partition by dining_segment order by year__month) as lag_ordercount,
    tot_revenue,
    lag(tot_revenue) over(partition by dining_segment order by year__month) as lag_tot_revenue
  from  dining_segment_summary 
  where weekpart = 'weekday')

select
  year__month,
  dining_segment,
  ordercount,
  round(((ordercount - lag_ordercount) / lag_ordercount)*100, 2) as ordercount_change_in_percent,
  round(tot_revenue, 2) as tot_revenue,
  round(((tot_revenue - lag_tot_revenue) / lag_tot_revenue)*100, 2) as tot_revenue_change_in_percent
from weekdaylag_dining_segment_summary;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
24 rows affected.


[('2022-6', '1-Breakfast', 1215, None, 33230.7, None),
 ('2022-7', '1-Breakfast', 1144, Decimal('-5.84'), 33229.6, -0.0),
 ('2022-8', '1-Breakfast', 1018, Decimal('-11.01'), 27481.0, -17.3),
 ('2022-9', '1-Breakfast', 998, Decimal('-1.96'), 27242.2, -0.87),
 ('2022-6', '2-Brunch', 806, None, 22517.5, None),
 ('2022-7', '2-Brunch', 690, Decimal('-14.39'), 19289.9, -14.33),
 ('2022-8', '2-Brunch', 650, Decimal('-5.80'), 17782.7, -7.81),
 ('2022-9', '2-Brunch', 574, Decimal('-11.69'), 15686.4, -11.79),
 ('2022-6', '3-Lunch', 1228, None, 33855.7, None),
 ('2022-7', '3-Lunch', 1129, Decimal('-8.06'), 31745.8, -6.23),
 ('2022-8', '3-Lunch', 996, Decimal('-11.78'), 28130.3, -11.39),
 ('2022-9', '3-Lunch', 973, Decimal('-2.31'), 27138.9, -3.52),
 ('2022-6', '4-Snacks', 1656, None, 46596.2, None),
 ('2022-7', '4-Snacks', 1500, Decimal('-9.42'), 41610.6, -10.7),
 ('2022-8', '4-Snacks', 1380, Decimal('-8.00'), 38004.8, -8.67),
 ('2022-9', '4-Snacks', 1173, Decimal('-15.00'), 32313.7, -14.97),
 ('2022-6', '5-Dinner', 1282, None, 35413.5, None),
 ('2022-7', '5-Dinner', 1088, Decimal('-15.13'), 29802.7, -15.84),
 ('2022-8', '5-Dinner', 992, Decimal('-8.82'), 28083.4, -5.77),
 ('2022-9', '5-Dinner', 925, Decimal('-6.75'), 25717.6, -8.42),
 ('2022-6', '6-After hours', 2856, None, 79044.1, None),
 ('2022-7', '6-After hours', 2545, Decimal('-10.89'), 69314.8, -12.31),
 ('2022-8', '6-After hours', 2348, Decimal('-7.74'), 66104.0, -4.63),
 ('2022-9', '6-After hours', 2246, Decimal('-4.34'), 61087.0, -7.59)]

#### <b> observations on weekdays </b>
* In the Breakfast and Lunch segments, the rate of decline accelerated between July and August, but slowed during June to July and August to September.
* In the Brunch and Snacks segments, the rate of decline slowed down from July to August, but accelerated during June to July and from August to September.
* In the Dinner and After-hours segments, the decline rate in order count gradually slowed over the months, whereas total revenue rate remained volatile.

In [22]:
%%sql

with weekendlag_dining_segment_summary as (
  select 
    year__month,
    dining_segment,
    ordercount,
    lag(ordercount) over(partition by dining_segment order by year__month) as lag_ordercount,
    tot_revenue,
    lag(tot_revenue) over(partition by dining_segment order by year__month) as lag_tot_revenue
  from  dining_segment_summary 
  where weekpart = 'weekend')

select
  year__month,
  dining_segment,
  ordercount,
  round(((ordercount - lag_ordercount) / lag_ordercount)*100, 2) as ordercount_change_in_percent,
  round(tot_revenue, 2) as tot_revenue,
  round(((tot_revenue - lag_tot_revenue) / lag_tot_revenue)*100, 2) as tot_revenue_change_in_percent
from weekendlag_dining_segment_summary;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
24 rows affected.


[('2022-6', '1-Breakfast', 485, None, 13627.1, None),
 ('2022-7', '1-Breakfast', 423, Decimal('-12.78'), 11492.2, -15.67),
 ('2022-8', '1-Breakfast', 352, Decimal('-16.78'), 10024.3, -12.77),
 ('2022-9', '1-Breakfast', 355, Decimal('0.85'), 10666.5, 6.41),
 ('2022-6', '2-Brunch', 322, None, 9342.3, None),
 ('2022-7', '2-Brunch', 264, Decimal('-18.01'), 7157.6, -23.39),
 ('2022-8', '2-Brunch', 246, Decimal('-6.82'), 7096.9, -0.85),
 ('2022-9', '2-Brunch', 213, Decimal('-13.41'), 5891.5, -16.98),
 ('2022-6', '3-Lunch', 472, None, 13442.3, None),
 ('2022-7', '3-Lunch', 436, Decimal('-7.63'), 12517.6, -6.88),
 ('2022-8', '3-Lunch', 396, Decimal('-9.17'), 11939.8, -4.62),
 ('2022-9', '3-Lunch', 320, Decimal('-19.19'), 8782.2, -26.45),
 ('2022-6', '4-Snacks', 603, None, 16235.5, None),
 ('2022-7', '4-Snacks', 559, Decimal('-7.30'), 15048.0, -7.31),
 ('2022-8', '4-Snacks', 508, Decimal('-9.12'), 14322.3, -4.82),
 ('2022-9', '4-Snacks', 418, Decimal('-17.72'), 11247.4, -21.47),
 ('2022-6', '5-Dinner', 467, None, 12602.6, None),
 ('2022-7', '5-Dinner', 409, Decimal('-12.42'), 11082.2, -12.06),
 ('2022-8', '5-Dinner', 373, Decimal('-8.80'), 9763.9, -11.9),
 ('2022-9', '5-Dinner', 354, Decimal('-5.09'), 10086.2, 3.3),
 ('2022-6', '6-After hours', 1110, None, 31670.0, None),
 ('2022-7', '6-After hours', 957, Decimal('-13.78'), 26310.5, -16.92),
 ('2022-8', '6-After hours', 848, Decimal('-11.39'), 24632.5, -6.38),
 ('2022-9', '6-After hours', 816, Decimal('-3.77'), 22301.5, -9.46)]

#### <b> observations on weekends</b>
* On Weekends, in the Snacks & Lunch time segment, the rate of decline in both order count and total revenue is accelerating over the months and requires more attention.
* On Weekends, in the Dinner segment, the rate of decline in both order count and total revenue is slowing down over the months. Revenue started rising in September, showing a 3.3% increase compared to the previous month (August).
* On Weekends, in the Breakfast segment, the rate of decline in order count appears to be fluctuating (volatility) but seems slowing down, while the rate of change in total revenue is slowing down. There was a 6.41% increase in total revenue in September compared to the previous month (August).
* In the After-hours segment, the rate of decline in total revenue is fluctuating, showing volatility, but it appears to be slowing down. Similarly, the rate of decline in order count is also slowing down without any fluctuations.

### <b> Impact of Discounts on Order Count and Total Revenue </b>

In [23]:
%%sql
select
month(order_date) as month_no,
round(avg(discount), 2) as avg_discount,
round(sum(discount) / sum(final_price)* 100, 3) as discount_ratio_,
count(distinct order_id) as ordercount,
round(sum(final_price), 2) as tot_revenue
from orders
group by month(order_date);

 * mysql+mysqlconnector://root:***@localhost/foodhunter
4 rows affected.


[(6, 3.82, 13.751, 12502, 347577.5),
 (7, 3.74, 13.522, 11144, 308601.5),
 (8, 3.87, 13.813, 10107, 283365.9),
 (9, 3.74, 13.562, 9365, 258161.1)]

#### <b> observation</b>
* It appears that offering discounts does not significantly influence total revenue or the number of orders.

### <b> Avg delivery time & Avg order rating & driver rating</b>

In [24]:
%%sql

drop view if exists avg_deliverytime_and_revenue;

create view avg_deliverytime_and_revenue as
  select
    order_id,
    order_date,
    customer_id,
    o.driver_id,
    timediff(delivered_time, order_time) as delivery_duration,
    delivery_fee,
    order_rating,
    rating as driver_rating
  from orders o join drivers d on o.driver_id = d.driver_id;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
0 rows affected.
0 rows affected.


[]

In [25]:
%%sql

with month_summary as (
  select
  *
  from 
    (select
      month(order_date) as month_no,
      avg(delivery_duration) as avg_delivery_duration,
      round(avg(order_rating), 2) as avg_order_rating,
      count(distinct order_id) as ordercount
    from avg_deliverytime_and_revenue
    group by month(order_date)) q1
),
month_lagsummary as(
select
month_no,
avg_delivery_duration,
lag(avg_delivery_duration) over(order by month_no) as lag_avg_delivery_duration,
avg_order_rating,
lag(avg_order_rating) over(order by month_no) as lag_order_rating,
ordercount,
lag(ordercount) over(order by month_no) as lag_ordercount
from
month_summary
)
select
month_no,
avg_delivery_duration,
round((avg_delivery_duration - lag_avg_delivery_duration) / (lag_avg_delivery_duration), 2) as delivery_duration_change_percent,
avg_order_rating,
round((avg_order_rating - lag_order_rating) / (lag_order_rating), 2) as order_rating_change_percent,
ordercount,
round((ordercount - lag_ordercount) / (lag_ordercount), 2) as ordercount_change_percent
from
month_lagsummary;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
4 rows affected.


[(6, Decimal('1915.1336'), None, 4.22, None, 12502, None),
 (7, Decimal('2114.2319'), Decimal('0.10'), 3.72, -0.12, 11144, Decimal('-0.11')),
 (8, Decimal('2413.0108'), Decimal('0.14'), 3.38, -0.09, 10107, Decimal('-0.09')),
 (9, Decimal('3111.8420'), Decimal('0.29'), 3.0, -0.11, 9365, Decimal('-0.07'))]

In [26]:
%%sql
# Identifying the top 10 worst-performing delivery partners with higher average delivery duration, lower order ratings and with poor driver ratings.
select
  driver_id,
  avg(delivery_duration) as avg_delivery_duration,
  round(avg(driver_rating), 2) as driver_rating,
  round(avg(order_rating), 2) as avg_order_rating
from avg_deliverytime_and_revenue
group by driver_id
order by avg_delivery_duration desc, avg_order_rating asc, driver_rating asc
limit 10;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
10 rows affected.


[(163, Decimal('2472.6316'), 4.0, 3.53),
 (93, Decimal('2455.5556'), 4.9, 3.55),
 (141, Decimal('2453.8462'), None, 3.35),
 (213, Decimal('2434.9206'), 4.6, 3.54),
 (27, Decimal('2427.2222'), 4.6, 3.74),
 (249, Decimal('2425.4438'), 4.0, 3.56),
 (250, Decimal('2425.4237'), 4.6, 3.59),
 (238, Decimal('2423.5632'), 4.0, 3.6),
 (57, Decimal('2421.7143'), 4.0, 3.68),
 (115, Decimal('2415.1515'), 3.7, 3.49)]

In [27]:
%%sql
#Identifying the top 10 best-performing delivery partners with lower average delivery duration, higher order ratings and better driver ratings.
select
  driver_id,
  avg(delivery_duration) as avg_delivery_duration,
  round(avg(driver_rating), 2) as driver_rating,
  round(avg(order_rating), 2) as avg_order_rating
from avg_deliverytime_and_revenue
group by driver_id
order by avg_delivery_duration asc, driver_rating desc, avg_order_rating desc
limit 10;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
10 rows affected.


[(244, Decimal('2239.0863'), 4.3, 3.7),
 (192, Decimal('2241.7647'), 3.2, 3.69),
 (37, Decimal('2244.8276'), 3.8, 3.67),
 (164, Decimal('2256.7251'), 3.7, 3.6),
 (22, Decimal('2256.9832'), 4.6, 3.79),
 (7, Decimal('2257.2289'), None, 3.63),
 (167, Decimal('2260.5128'), 5.0, 3.64),
 (128, Decimal('2261.7647'), 3.1, 3.67),
 (219, Decimal('2269.7183'), 3.8, 3.59),
 (133, Decimal('2270.7317'), 4.5, 3.75)]

#### <b> observation</b>
* We observe a significant inverse relationship between delivery duration and order count as well as order rating. This suggests that delivery duration impacts revenue, order count and order rating.
* Over the four-month period, as delivery duration increases, order rating, number of orders and total revenue decrease accordingly.

### <b> Food Items preferences</b>

In [28]:
%%sql

drop view if exists foodhunter.fooditems_preferences;

create view fooditems_preferences as
select
*
from orders o
join orders_items oi using(order_id)
join food_items fi using(item_id)
join restaurants r using(restaurant_id);

 * mysql+mysqlconnector://root:***@localhost/foodhunter
0 rows affected.
0 rows affected.


[]

In [29]:
%%sql

select
food_type, count(distinct order_id) as ordercount, sum(quantity) as quantity_sum, count(distinct item_id) as item_count, count(distinct cuisine) as cuisine_count
from fooditems_preferences
group by food_type;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
2 rows affected.


[('non-veg', 34664, Decimal('63077'), 219, 4),
 ('veg', 15454, Decimal('23239'), 81, 4)]

In [30]:
%%sql

update food_items 
set food_type = case
                 when food_type like 'v%' then 'veg'
                 when food_type like 'non-%' then 'non-veg'
                 else 'others'
end;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
314 rows affected.


[]

In [31]:
%%sql

select
food_type, count(distinct order_id) as ordercount, sum(quantity) as quantity_sum, count(distinct item_id) as item_count, count(distinct cuisine) as cuisine_count
from fooditems_preferences
group by food_type;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
2 rows affected.


[('non-veg', 34664, Decimal('63077'), 219, 4),
 ('veg', 15454, Decimal('23239'), 81, 4)]

In [32]:
%%sql
#Items that have not been ordered by the customers over the 4 months from June to September
select
item_id, item_name, food_type, restaurant_id, restaurant_name, cuisine, price
from orders_items
right join food_items using(item_id)
join restaurants using(restaurant_id)
where order_id is null;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
14 rows affected.


[(414, 'Pepperoni pizza', 'non-veg', 99, 'Egg Shop', 'Indian', None),
 (412, 'Bellini', 'veg', 58, 'Le Coucou', 'Italian', None),
 (406, 'Special Pasta', 'non-veg', 13, "Roberta's Pizza", 'Italian', None),
 (407, 'Pepperoni pizza', 'non-veg', 13, "Roberta's Pizza", 'Italian', None),
 (408, 'Papadum', 'veg', 16, 'La Vara', 'Italian', None),
 (411, 'Special Pasta', 'non-veg', 58, 'Le Coucou', 'Italian', None),
 (413, 'Special Pasta', 'non-veg', 60, 'Frenchette', 'Italian', None),
 (403, 'Nam Tok', 'non-veg', 4, 'Minetta Tavern', 'Thai', None),
 (405, 'Tod Man Pla', 'non-veg', 75, 'Pam Real Thai', 'Thai', None),
 (409, 'Fiano di Avellino', 'veg', 35, 'Buvette', 'Italian', None),
 (402, 'Kana Moo Krob', 'non-veg', 4, 'Minetta Tavern', 'Thai', None),
 (401, 'Pad Sator', 'non-veg', 4, 'Minetta Tavern', 'Thai', None),
 (404, 'Kana Moo Krob', 'non-veg', 75, 'Pam Real Thai', 'Thai', None),
 (410, 'Cozze al vino bianco', 'non-veg', 35, 'Buvette', 'Italian', None)]

#### <b> observations</b>
* Non-veg items are preferred over veg items.
* We should consider expanding the variety of non-vegetarian food items and introducing a wider range of regional and international non-veg cuisines to attract more customers.
* At the same time, we can adopt a unique marketing strategy for vegetarian dishes by reintroducing traditional and culturally rich recipes — especially those that date back 5–6 centuries, whether from local heritage or other fellow cultures. This approach will create curiosity and appeal for customers looking to try something new, authentic and deeply rooted in history.
* We can observe that most of the dishes not preferred by people over the four months from June to September are Italian (8 out of 14 dishes), followed by Thai dishes (5 out of 14) and then only one Indian dish.
* The following restaurants have not received any orders through our FoodHunter app: <b><i> Egg Shop, Le Coucou, Roberta's Pizza, La Vara, Frenchette, Minetta Tavern, Pam Real Thai and Buvette.</i> </b>
* Non-vegetarian dishes hold a larger share than vegetarian dishes in the not preferred list.

### <b> customer demographics analysis</b>

In [33]:
%%sql
#checking for mismatches in addresses
select
count(*) address_mismatch
from orders o
join customers c on o.customer_id = c.customer_id
where o.delivery_address <> c.address;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
1 rows affected.


[(0,)]

In [34]:
%%sql
#top-performing customer locations
select
place_name,
round(sum(final_price), 2) as tot_revenue,
count(distinct order_id) as ordercount,
count(distinct customer_id) as customer_count
from
  (select
     o.order_id,
     o.final_price,
     c.customer_id, c.address,
     substring(c.address, instr(c.address, ' '), length(c.address)) as place_name
   from orders o
   inner join customers c on o.customer_id = c.customer_id) q1
group by place_name
order by tot_revenue desc, ordercount desc, customer_count desc, place_name asc
limit 10;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
10 rows affected.


[(' Parkview Avenue', 10890.4, 343, 73),
 (' Oak Street', 10807.2, 364, 77),
 (' Greenfield Terrace', 10781.1, 368, 77),
 (' Washington Avenue', 10674.9, 354, 74),
 (' Chestnut Avenue', 10491.0, 376, 75),
 (' Beechwood Road', 10430.9, 357, 75),
 (' High Street', 10206.3, 345, 76),
 (' Glenwood Avenue', 10194.9, 333, 76),
 (' Northview Road', 10152.0, 376, 75),
 (' Brookdale Place', 10108.4, 339, 76)]

In [35]:
%%sql
#lowest-performing customer locations 
select
place_name,
round(sum(final_price), 2) as tot_revenue,
count(distinct order_id) as ordercount,
count(distinct customer_id) as customer_count
from
  (select
     o.order_id,
     o.final_price,
     c.customer_id, c.address,
     substring(c.address, instr(c.address, ' '), length(c.address)) as place_name
   from orders o
   inner join customers c on o.customer_id = c.customer_id) q1
group by place_name
order by tot_revenue asc, ordercount asc, customer_count asc, place_name asc
limit 10;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
10 rows affected.


[(' Belmont Avenue', 7115.1, 262, 73),
 (' Orchard Road', 7267.0, 303, 75),
 (' Holly Lane', 7416.2, 292, 69),
 (' Beech Street', 7437.0, 292, 71),
 (' Yorktown Road', 7502.8, 294, 71),
 (' Northview Terrace', 7601.8, 301, 73),
 (' Brookside Avenue', 7608.2, 285, 73),
 (' Pinehurst Drive', 7766.1, 283, 70),
 (' Valley View Road', 7778.2, 287, 71),
 (' Maplewood Road', 7812.7, 305, 72)]

In [36]:
%%sql
#places with inactive orders over the 4 Months
select
place_name,
count(distinct customer_id) as customer_count
from
  (select
     c.customer_id, c.address,
     substring(c.address, instr(c.address, ' '), length(c.address)) as place_name
   from orders o
   right join customers c on o.customer_id = c.customer_id
   where o.order_id is null) q1
group by place_name
order by customer_count desc, place_name
limit 20;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
20 rows affected.


[(' Fairview Avenue', 4),
 (' Oak Hill Drive', 4),
 (' Springdale Road', 4),
 (' Briarwood Court', 3),
 (' Cedar Street', 3),
 (' Elmwood Road', 3),
 (' Grove Street', 3),
 (' Longview Road', 3),
 (' Maplewood Road', 3),
 (' Northview Terrace', 3),
 (' Pine Street', 3),
 (' Rosedale Avenue', 3),
 (' Yorktown Road', 3),
 (' Ashland Avenue', 2),
 (' Beech Street', 2),
 (' Broad Street', 2),
 (' Brookside Avenue', 2),
 (' Cedar Terrace', 2),
 (' Eastlake Drive', 2),
 (' Fairway Drive', 2)]

#### <b> observations</b>
* <b> Top Performing Customer Locations (by revenue, orders, and customer count): </b>
<i> Parkview Avenue > Oak Street > Greenfield Terrace > Washington Avenue > Chestnut Avenue > Beechwood Road > High Street > Glenwood Avenue > Northview Road > Brookdale Place. </i>
*  <b> Lowest Performing Customer Locations (by revenue, orders, and customer count): </b>
<i> Belmont Avenue < Orchard Road < Holly Lane < Beech Street < Yorktown Road < Northview Terrace < Brookside Avenue < Pinehurst Drive < Valley View Road < Maplewood Road. </i>
* <b> Top 10 Places with Inactive orders over the 4 Months (by customer base): </b>
<i> Fairview Avenue > Oak Hill Drive > Springdale Road > Briarwood Court > Cedar Street > Elmwood Road > Grove Street > Longview Road > Maplewood Road > Northview Terrace </i>

In [6]:
%%sql

select
cat,
count(distinct order_id) as ordercount
from
(select
*,
case when total_price between 3 and 50 then "a"
when total_price between 51 and 100 then "b"
when total_price between 101 and 150 then "c"
end as cat
from orders) q1
group by cat;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
3 rows affected.


[('a', 38312), ('b', 3118), ('c', 1688)]

In [5]:
%%sql

select
cat,
count(distinct order_id) as ordercount
from
(select
*,
case when total_price between 3 and 40 then "a"
when total_price between 41 and 80 then "b"
when total_price between 81 and 150 then "c"
end as cat
from orders) q1
group by cat;

 * mysql+mysqlconnector://root:***@localhost/foodhunter
3 rows affected.


[('a', 35486), ('b', 4592), ('c', 3040)]